File: services/copy.py
import pyodbc
import psycopg2
import schedule
import time

# Database connection settings
mssql_conn_str = 'DRIVER={ODBC Driver 17 for SQL Server};SERVER=127.0.0.1;DATABASE=biotime;UID=Digitali;PWD=Digitali'
pg_conn_str = 'dbname=decan user=postgres password=password123 host=192.168.0.108 port=5432'

def fetch_latest_id_from_postgres():
    with psycopg2.connect(pg_conn_str) as pg_conn:
        with pg_conn.cursor() as cursor:
            cursor.execute('SELECT MAX(id) FROM public.logs')
            result = cursor.fetchone()
            return result[0] if result[0] is not None else 0

def fetch_new_records_from_mssql(latest_id):
    with pyodbc.connect(mssql_conn_str) as mssql_conn:
        query = 'SELECT [id], [employeeid], [direction], [shortname], [serialno], [log_datetime] FROM [biotime].[dbo].[logs] WHERE [id] > ?'
        with mssql_conn.cursor() as cursor:
            cursor.execute(query, latest_id)
            return cursor.fetchall()

def insert_records_into_postgres(records):
    with psycopg2.connect(pg_conn_str) as pg_conn:
        with pg_conn.cursor() as cursor:
            insert_query = '''
            INSERT INTO public.logs (id, employeeid, direction, shortname, serialno, log_datetime)
            VALUES (%s, %s, %s, %s, %s, %s)
            '''
            cursor.executemany(insert_query, records)
            pg_conn.commit()

def job():
    latest_id = fetch_latest_id_from_postgres()
    new_records = fetch_new_records_from_mssql(latest_id)
    
    if new_records:
        insert_records_into_postgres(new_records)
        print(f'Inserted {len(new_records)} new records into PostgreSQL.')
    else:
        print('No new records to insert.')

# Schedule the job to run every 10 seconds
schedule.every(10).seconds.do(job)

while True:
    schedule.run_pending()
    time.sleep(1)

File: services/copy.go
package main

import (
	"database/sql"
	"fmt"
	"log"
	"time"

	_ "github.com/denisenkom/go-mssqldb"
	_ "github.com/lib/pq"
)

// Database connection settings
const (
	mssqlConnStr = "server=127.0.0.1;user id=Digitali;password=Digitali;database=biotime"
	pgConnStr    = "host=10.38.21.181 port=5432 user=postgres password=password123 dbname=casa sslmode=disable"
)

func fetchLatestIDFromPostgres() (int, error) {
	db, err := sql.Open("postgres", pgConnStr)
	if err != nil {
		return 0, err
	}
	defer db.Close()

	var latestID int
	err = db.QueryRow("SELECT COALESCE(MAX(id), 0) FROM public.logs").Scan(&latestID)
	if err != nil {
		return 0, err
	}
	return latestID, nil
}

func fetchNewRecordsFromMSSQL(latestID int) ([][6]interface{}, error) {
	db, err := sql.Open("sqlserver", mssqlConnStr)
	if err != nil {
		return nil, err
	}
	defer db.Close()

	query := `SELECT [id], [employeeid], [direction], [shortname], [serialno], [log_datetime] 
			  FROM [biotime].[dbo].[logs] WHERE [id] > @latestID`

	rows, err := db.Query(query, sql.Named("latestID", latestID))
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var records [][6]interface{}
	for rows.Next() {
		var record [6]interface{}
		err := rows.Scan(&record[0], &record[1], &record[2], &record[3], &record[4], &record[5])
		if err != nil {
			return nil, err
		}
		records = append(records, record)
	}
	return records, nil
}

func insertRecordsIntoPostgres(records [][6]interface{}) error {
	db, err := sql.Open("postgres", pgConnStr)
	if err != nil {
		return err
	}
	defer db.Close()

	insertQuery := `
		INSERT INTO public.logs (id, employeeid, direction, shortname, serialno, log_datetime)
		VALUES ($1, $2, $3, $4, $5, $6)`

	tx, err := db.Begin()
	if err != nil {
		return err
	}

	stmt, err := tx.Prepare(insertQuery)
	if err != nil {
		tx.Rollback()
		return err
	}
	defer stmt.Close()

	for _, record := range records {
		_, err := stmt.Exec(record[0], record[1], record[2], record[3], record[4], record[5])
		if err != nil {
			tx.Rollback()
			return err
		}
	}

	err = tx.Commit()
	if err != nil {
		return err
	}
	return nil
}

func job() {
	latestID, err := fetchLatestIDFromPostgres()
	if err != nil {
		log.Printf("Error fetching latest ID: %v\n", err)
		return
	}

	newRecords, err := fetchNewRecordsFromMSSQL(latestID)
	if err != nil {
		log.Printf("Error fetching new records from MSSQL: %v\n", err)
		return
	}

	if len(newRecords) > 0 {
		err = insertRecordsIntoPostgres(newRecords)
		if err != nil {
			log.Printf("Error inserting records into PostgreSQL: %v\n", err)
			return
		}
		fmt.Printf("Inserted %d new records into PostgreSQL.\n", len(newRecords))
	} else {
		fmt.Println("No new records to insert.")
	}
}

func main() {
	for {
		job()
		time.Sleep(10 * time.Second)
	}
}
